src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1721 @@
+        let (g_0, g_1, g) = {
+            let g_0 = rho_val.map(|rho| bitrange_subset(rho, 254..255));
+            let g_1 = psi_val.map(|psi| bitrange_subset(psi, 0..9));
+            let g_2 = psi_val.map(|psi| bitrange_subset(psi, 9..249));
+
+            // Constrain g_1 to be 9 bits.
+            let g_1 = self.sinsemilla_config.lookup_config.witness_short_check(
+                layouter.namespace(|| "g_1 is 9 bits"),
+                g_1,
+                9,
+            )?;
+
+            // g_0 will be boolean-constrained in the gate.
+            // g_2 = z1_g from the SinsemillaHash(g) running sum output.
+
+            let g = g_0.zip(g_1.value()).zip(g_2).map(|((g_0, g_1), g_2)| {
+                let g1_shifted = g_1 * pallas::Base::from_u64(2);
+                let g2_shifted = g_2 * pallas::Base::from_u64(1 << 10);
+                g_0 + g1_shifted + g2_shifted
+            });
+            let g = MessagePiece::from_field_elem(chip.clone(), layouter.namespace(|| "g"), g, 25)?;
+
+            (g_0, g_1, g)
+        };
+
+        // h = h_0 || h_1 || h_2
+        //   = (bits 249..=253 of psi) || (bit 254 of psi) || 4 zero bits
+        let (h_0, h_1, h) = {
+            let h_0 = psi_val.map(|psi| bitrange_subset(psi, 249..254));
+            let h_1 = psi_val.map(|psi| bitrange_subset(psi, 254..255));
+
+            // Constrain h_0 to be 5 bits.
+            let h_0 = self.sinsemilla_config.lookup_config.witness_short_check(
+                layouter.namespace(|| "h_0 is 5 bits"),
+                h_0,
+                5,
+            )?;
+
+            // h_1 will be boolean-constrained in the gate.
+
+            let h = h_0
+                .value()
+                .zip(h_1)
+                .map(|(h_0, h_1)| h_0 + h_1 * pallas::Base::from_u64(1 << 5));
+            let h = MessagePiece::from_field_elem(chip.clone(), layouter.namespace(|| "h"), h, 1)?;
+
+            (h_0, h_1, h)
+        };
+
+        // Check decomposition of `y(g_d)`.
+        let b_2 = self.y_canonicity(layouter.namespace(|| "y(g_d) decomposition"), g_d.y(), b_2)?;
+        // Check decomposition of `y(pk_d)`.
+        let d_1 = self.y_canonicity(
+            layouter.namespace(|| "y(pk_d) decomposition"),
+            pk_d.y(),
+            d_1,
+        )?;
+
+        let (cm, zs) = {
+            let message = Message::from_pieces(
+                chip.clone(),
+                vec![
+                    a.clone(),
+                    b.clone(),
+                    c.clone(),
+                    d.clone(),
+                    e.clone(),
+                    f.clone(),
+                    g.clone(),
+                    h.clone(),
+                ],
+            );
+            let domain = CommitDomain::new(chip, ecc_chip, &SinsemillaCommitDomains::NoteCommit);
+            domain.commit(
+                layouter.namespace(|| "Process NoteCommit inputs"),
+                message,
+                rcm,
+            )?
```

Comment:
Yeah, I think it would be possible to automate this kind of processing. That's clearly necessary to make the Sinsemilla gadget accessible to a wider user base.

We also probably didn't do ourselves any favours by specifying bit-packed inputs to Sinsemilla, rather than a more regular input arrangement of FE(x(g<sub>d</sub>)) || FE(x(pk<sub>d</sub>) || FE(ρ) || FE(ψ) || FE(v + 2<sup>64</sup> ỹ(g<sub>d</sub>) + 2<sup>65</sup> ỹ(pk<sub>d</sub>)).